src/server/engine_shard_set.cc

```diff
@@ -605,6 +605,11 @@ void EngineShard::Heartbeat() {
   if (IsReplica())  // Never run expiration on replica.
     return;
 
+  // TODO: iterate over all namespaces
+  DbSlice& db_slice = namespaces.GetDefaultNamespace().GetDbSlice(shard_id());
+  // Some of the functions below might acquire the lock again so we need to unlock it
+  std::unique_lock lk(db_slice.GetSerializationMutex());
```

Comment:
Are we sure that `ABSL_LOCKS_EXCLUDED` work as intended ? Because I injected a deadlock here that violated the invariant of `ABSL_LOCKS_EXLUDED` and it deadlocked spectacularly (neither a compiler or runtime error). From a quick the macros used `__attribute__` but IMO I did not see any safeguard in practice which makes me question if I did something wrong or if those are not really that useful.

Reply:
I guess I will need to do some further reading when I get a chance. Two important links:

1. https://releases.llvm.org/3.5.0/tools/clang/docs/ThreadSafetyAnalysis.html
2. https://releases.llvm.org/3.5.0/tools/clang/docs/ThreadSafetyAnalysis.html#limitations

Reply:
I need to check if we use  `-Wthread-safety` because otherwise the analysis is not run. Also it's static_analysis (I was also concerned if this had any runtime impact)

Reply:
@kostasrim the `ABSL_*` checks are compile time only.
They are only enabled in clang, are you using that locally?

Reply:
discussed internally. attaching the issue for completeness https://github.com/dragonflydb/dragonfly/issues/3448